tree_sitter_grep/src/plugin.rs

```diff
@@ -0,0 +1,80 @@
+use libloading::Library;
+use once_cell::sync::OnceCell;
+use std::ffi::OsStr;
+use tree_sitter::Node;
+use tree_sitter_grep_plugin_core::Filterer;
+use tree_sitter_grep_plugin_core::PluginDeclaration;
+
+use std::sync::Arc;
+
+pub struct FiltererProxy {
+    filterer: Box<dyn Filterer>,
+    _lib: Arc<Library>,
+}
+
+impl Filterer for FiltererProxy {
+    fn call(&self, node: &Node) -> bool {
+        self.filterer.call(node)
+    }
+}
+
+struct PluginRegistrar {
+    filterer: Option<FiltererProxy>,
+    lib: Arc<Library>,
+}
+
+impl PluginRegistrar {
+    pub fn new(lib: Arc<Library>) -> Self {
+        Self {
+            lib,
+            filterer: Default::default(),
+        }
+    }
+}
+
+impl tree_sitter_grep_plugin_core::PluginRegistrar for PluginRegistrar {
+    fn register_filterer(&mut self, filterer: Box<dyn Filterer>) {
+        let proxy = FiltererProxy {
+            filterer,
+            _lib: self.lib.clone(),
+        };
+        self.filterer = Some(proxy);
+    }
+}
+
+fn load_plugin(library_path: impl AsRef<OsStr>, filter_arg: Option<&str>) -> FiltererProxy {
+    let library = Arc::new(unsafe {
+        Library::new(library_path).expect("Couldn't load filter dynamic library")
+    });
+
+    let plugin_declaration = unsafe {
+        library
+            .get::<*mut PluginDeclaration>(b"plugin_declaration\0")
```

Comment:
Ok so this is again mimicking that tutorial. Other things I was reading were [this](https://nullderef.com/blog/plugin-start/) and the corresponding [Reddit discussion](https://www.reddit.com/r/rust/comments/q2n6b8/plugins_in_rust_diving_into_dynamic_loading/) - they warn that "Rust-to-Rust FFI" really technically isn't safe at all (even with sanity-checking that you're using the same version of the Rust compiler etc at dynamic-library-loading time) because the Rust compiler reserves the right to eg change struct layouts _between compiler runs_, but that it sounds like (at least currently) it never really does that in practice

That series looks helpful and well-written in general, it's doing some exploring/explaining of alternatives eg WASM